File: packages/liftoff/android/src/main/java/com/margelo/nitro/liftoffads/HybridLiftoffAds.kt
package com.margelo.nitro.liftoffads

import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.facebook.react.bridge.ReactApplicationContext
import com.margelo.nitro.NitroModules
import com.margelo.nitro.core.Promise
import com.vungle.ads.InitializationListener
import com.vungle.ads.VungleAds
import com.vungle.ads.VungleError
import com.vungle.ads.VunglePrivacySettings
import kotlinx.coroutines.CompletableDeferred
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.withContext

@DoNotStrip
@Keep
class HybridLiftoffAds : HybridLiftoffAdsSpec() {
  private val context: ReactApplicationContext by lazy {
    NitroModules.applicationContext as? ReactApplicationContext
      ?: error("Liftoff Ads requires a ReactApplicationContext")
  }

  private val stateLock = Any()
  private var initializedAppId: String? = null
  private var initializingAppId: String? = null
  private var initialization: CompletableDeferred<Unit>? = null

  override fun initialize(appId: String): Promise<Unit> = Promise.async {
    require(appId.isNotBlank()) { "Liftoff app ID is empty" }

    val alreadyInitialized = synchronized(stateLock) {
      if (VungleAds.isInitialized()) {
        val existing = initializedAppId
        if (existing != null && existing != appId) {
          error("Liftoff SDK was already initialized with another app ID")
        }
        initializedAppId = appId
        true
      } else {
        false
      }
    }
    if (alreadyInitialized) return@async

    val (deferred, shouldStart) = synchronized(stateLock) {
      val current = initialization
      if (current != null) {
        if (initializingAppId != appId) {
          error("Liftoff SDK is being initialized with another app ID")
        }
        current to false
      } else {
        CompletableDeferred<Unit>().also {
          initialization = it
          initializingAppId = appId
        } to true
      }
    }

    if (shouldStart) {
      withContext(Dispatchers.Main) {
        VungleAds.init(
          context.applicationContext,
          appId,
          object : InitializationListener {
            override fun onSuccess() {
              synchronized(stateLock) {
                initializedAppId = appId
                initialization = null
                initializingAppId = null
              }
              deferred.complete(Unit)
            }

            override fun onError(vungleError: VungleError) {
              synchronized(stateLock) {
                if (initialization === deferred) initialization = null
                initializingAppId = null
              }
              deferred.completeExceptionally(
                liftoffException("initialize", vungleError),
              )
            }
          },
        )
      }
    }

    deferred.await()
  }

  override fun setGDPRConsent(optIn: Boolean, consentMessageVersion: String) {
    VunglePrivacySettings.setGDPRStatus(optIn, consentMessageVersion)
  }

  override fun setCCPAConsent(optIn: Boolean) {
    VunglePrivacySettings.setCCPAStatus(optIn)
  }

  override fun setCOPPA(isUserCoppa: Boolean) {
    VunglePrivacySettings.setCOPPAStatus(isUserCoppa)
  }

  override fun createInterstitialAd(
    placementId: String,
  ): HybridLiftoffInterstitialAdSpec {
    requireInitialized(placementId)
    return HybridLiftoffInterstitialAd(context, placementId)
  }

  override fun createRewardedAd(placementId: String): HybridLiftoffRewardedAdSpec {
    requireInitialized(placementId)
    return HybridLiftoffRewardedAd(context, placementId)
  }

  private fun requireInitialized(placementId: String) {
    require(VungleAds.isInitialized()) { "Liftoff SDK is not initialized" }
    require(placementId.isNotBlank()) { "Liftoff placement ID is empty" }
  }
}

File: packages/liftoff/ios/HybridLiftoffAds.swift
import Foundation
import NitroModules
import VungleAdsSDK

final class HybridLiftoffAds: HybridLiftoffAdsSpec, @unchecked Sendable {
  private let stateLock = NSLock()
  private var initializedAppId: String?
  private var initializingAppId: String?
  private var initializationContinuations: [CheckedContinuation<Void, Error>] = []
  private var isInitializing = false

  func initialize(appId: String) throws -> Promise<Void> {
    guard !appId.isEmpty else {
      throw liftoffError("Liftoff app ID is empty")
    }

    return Promise.async {
      try await withCheckedThrowingContinuation { continuation in
        DispatchQueue.main.async {
          self.stateLock.lock()
          if VungleAds.isInitialized() {
            let existing = self.initializedAppId
            if existing == nil { self.initializedAppId = appId }
            self.stateLock.unlock()

            if let existing, existing != appId {
              continuation.resume(
                throwing: liftoffError("Liftoff SDK was initialized with another app ID")
              )
            } else {
              continuation.resume()
            }
            return
          }

          if self.isInitializing, self.initializingAppId != appId {
            self.stateLock.unlock()
            continuation.resume(
              throwing: liftoffError("Liftoff SDK is being initialized with another app ID")
            )
            return
          }

          self.initializationContinuations.append(continuation)
          if self.isInitializing {
            self.stateLock.unlock()
            return
          }

          self.isInitializing = true
          self.initializingAppId = appId
          self.stateLock.unlock()

          VungleAds.initWithAppId(appId) { error in
            self.stateLock.lock()
            let continuations = self.initializationContinuations
            self.initializationContinuations.removeAll()
            self.isInitializing = false
            self.initializingAppId = nil
            if error == nil { self.initializedAppId = appId }
            self.stateLock.unlock()

            continuations.forEach { item in
              if let error {
                item.resume(throwing: error)
              } else {
                item.resume()
              }
            }
          }
        }
      }
    }
  }

  func setGDPRConsent(optIn: Bool, consentMessageVersion: String) throws {
    runOnMain {
      VunglePrivacySettings.setGDPRStatus(optIn)
      VunglePrivacySettings.setGDPRMessageVersion(consentMessageVersion)
    }
  }

  func setCCPAConsent(optIn: Bool) throws {
    runOnMain { VunglePrivacySettings.setCCPAStatus(optIn) }
  }

  func setCOPPA(isUserCoppa: Bool) throws {
    runOnMain { VunglePrivacySettings.setCOPPAStatus(isUserCoppa) }
  }

  func createInterstitialAd(
    placementId: String
  ) throws -> any HybridLiftoffInterstitialAdSpec {
    try requireInitialized(placementId: placementId)
    return HybridLiftoffInterstitialAd(placementId: placementId)
  }

  func createRewardedAd(
    placementId: String
  ) throws -> any HybridLiftoffRewardedAdSpec {
    try requireInitialized(placementId: placementId)
    return HybridLiftoffRewardedAd(placementId: placementId)
  }

  private func requireInitialized(placementId: String) throws {
    guard VungleAds.isInitialized() else {
      throw liftoffError("Liftoff SDK is not initialized")
    }
    guard !placementId.isEmpty else {
      throw liftoffError("Liftoff placement ID is empty")
    }
  }
}
